### PR TITLE
Pca for sparse data

### PR DESCRIPTION
**Pull request by Koncopd**
For this - https://github.com/theislab/scanpy/issues/393

Benchmarks
https://github.com/Koncopd/anndata-scanpy-benchmarks/blob/master/pca_for_sparse.ipynb

### CODE REVIEW

**Comment by falexwolf** [comment]
This is nice! Thank you!

It appears to me that the benchmarks show that this only becomes relevant for *very* large data. So we need to be mindful to not break backward compatibility for all the small and medium-size datasets that people use (which we do by introducing the tiny difference). Don't you think that in the light of this, it would be better to leave the default as is (densifying) and have an option `sparse_pca` or something similar?

**Comment by Koncopd** [comment]
>It appears to me that the benchmarks show that this only becomes relevant for very large data.

Hm, even for my example it is 77.14 MiB vs 893.92 MiB, so 10 times difference. This seems large to me, no?

**Comment by falexwolf** [comment]
> Hm, even for my example it is 77.14 MiB vs 893.92 MiB, so 10 times difference. This seems large to me, no?

Yes, it's definitely large and it's awesome that you solved this problem! I just meant that it's not hitting people's computational resources limits: your example is 60K x 2K, so quite big already, if you densify you need 800MB, which is easily available even on a laptop. That's what I meant.

What do you think?

**Comment by Koncopd** [comment]
Yeah, this seems important only for large datasets in that sense. So, i will add sparse_pca option with False by default.

**Comment by Koncopd** [comment]
Not sure what kind of test to add for this...

**Comment by falexwolf** [comment]
As discussed, @Koncopd will try to integrate this into scikit-learn itself and not into Scanpy. :smile:

**Comment by Koncopd** [comment]
Similar pull request exists already in sklearn.
https://github.com/scikit-learn/scikit-learn/pull/12841
Will watch.

**Review by cyrus303 — APPROVED** [review]

**Comment by flying-sheep** [comment]
> So, i will add sparse_pca option with False by default.

How about make it smartly decide by default, informed by the benchmarks? You can describe in the param docs when it happens, and give a debug message which one was choosen.

File: scanpy/preprocessing/pca_for_sparse.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+
```
Comment (flying-sheep) [inline scanpy/preprocessing/pca_for_sparse.py:1]: File needs to be named starting with _ as the others

File: scanpy/preprocessing/pca_for_sparse.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import numpy as np
+from distutils.version import StrictVersion
```
Comment (flying-sheep) [inline scanpy/preprocessing/pca_for_sparse.py:3]: `distutils.version` is broken. Please use `from packaging import version` and then `version.parse`. (I added `packaging` to the requirements on master)

File: scanpy/preprocessing/pca_for_sparse.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+# if its dtype is float64
+# https://github.com/scikit-learn/scikit-learn/blob/f0ab589f1541b1ca4570177d93fd7979613497e3/sklearn/utils/sparsefuncs_fast.pyx#L72
+
+if StrictVersion(sklearn_version) < StrictVersion('0.20.0'):
+    from sklearn.utils.sparsefuncs_fast import _csr_mean_variance_axis0
+    mean_variance = lambda X: _csr_mean_variance_axis0(X.data, X.shape, X.indices)
+else:
+    from sklearn.utils.sparsefuncs_fast import csr_mean_variance_axis0 as mean_variance
+
+# need to pass issparse check
```
Comment (flying-sheep) [inline scanpy/preprocessing/pca_for_sparse.py:19]: I don’t understand this comment.

File: scanpy/preprocessing/pca_for_sparse.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        m, var = mean_variance(X)
+        m = np.asmatrix(m)
+        var = var.sum()
+        C = CentSparse(X, m)
+
+        if k >= n_features:
+            raise ValueError('n_components must be < n_features;'
+                             ' got %d >= %d' % (k, n_features))
+        U, Sigma, VT = randomized_svd(C, k, n_iter=self.n_iter,
```
Comment (flying-sheep) [inline scanpy/preprocessing/pca_for_sparse.py:63]: please don’t use uppercase letters in variable names.

File: scanpy/preprocessing/simple.py [hunk 5]
```diff
@@ -354,7 +354,9 @@ def pca(
     data: Union[AnnData, np.ndarray, spmatrix],
     n_comps: int = N_PCS,
     zero_center: Optional[bool] = True,
+    sparse_pca: Optional[bool] = False,
```
Comment (flying-sheep) [inline scanpy/preprocessing/simple.py:357]: Maybe make it `None` by default, and, based on the benchmarks, use it when the data is sufficiently large.

In this case, make sure you document this in the parameter docs and log a debug message.

**Review by flying-sheep — CHANGES_REQUESTED** [review]
Sorry about causing merge conflicts. Squashing all your commits and then rebasing on master is probably going to be easier than rebasing all the commits individually

**Comment by gokceneraslan** [comment]
Shall we replace sparse_pca with something else like implicit_centering or so? 

sparse_pca might give rise to confusion since there is indeed a PCA technique called sparse PCA.

**Comment by Koncopd** [comment]
Hm, it was decided to suspend this pr earlier.
There is an analogous pr in scikit-learn, but i'm not sure it will got forward.
I'm not sure what to do with this pr...

**Comment by flying-sheep** [comment]
Ah, then we should leave it open and unchanged until there’s a decision on the sklearn PR. Sorry I missed the decision.

**Comment by VolkerBergen** [comment]
What's the current state here?

This just came out 
https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1900-3
favoring sklearn's PCA implementation while stating that it cannot yet handle sparse matrices. Is that still true?

**Comment by Koncopd** [comment]
Last time i checked it was true. Need to check again, but i bet it is still the same.

**Comment by ivirshup** [comment]
That's still the case (at least for randomized PCA @Koncopd linked above), though it looks like there may be a another path forward using other solvers: https://github.com/scikit-learn/scikit-learn/issues/12794. Still needs an implementation though.

**Comment by atarashansky** [comment]
You could just add a `sparse` argument to `pca`. If True, just call this function instead of scikit-learn's PCA:

```
def sparse_pca(X,npcs,mu = None):
    # X -- scipy sparse data matrix
    # npcs -- number of principal components
    # mu -- precomputed feature means. if None, calculates them from X.

    # compute mean of data features
    if mu is None: 
        mu = X.mean(0).A.flatten()[None,:]

    # dot product operator for the means
    mmat = mdot = mu.dot 
    # dot product operator for the transposed means
    mhmat = mhdot = mu.T.dot 
    # dot product operator for the data
    Xmat = Xdot = X.dot 
    # dot product operator for the transposed data
    XHmat = XHdot = X.T.conj().dot 
    # dot product operator for a vector of ones
    ones = np.ones(X.shape[0])[None,:].dot 

    # modify the matrix/vector dot products to subtract the means
    def matvec(x): 
        return Xdot(x) - mdot(x)
    def matmat(x): 
        return Xmat(x) - mmat(x)
    def rmatvec(x): 
        return XHdot(x) - mhdot(ones(x))
    def rmatmat(x): 
        return XHmat(x) - mhmat(ones(x))
    
    # construct the LinearOperator
    XL = sp.sparse.linalg.LinearOperator(matvec = matvec, dtype = X.dtype,
                                         matmat = matmat,
                                         shape = X.shape,
                                        rmatvec = rmatvec, rmatmat = rmatmat)
     
    u,s,v = sp.sparse.linalg.svds(XL,solver='arpack',k=npcs)
    
    # i like my eigenvalues sorted in decreasing order
    idx = np.argsort(-s)
    S = np.diag(s[idx])
    # principal components
    pcs = u[:,idx].dot(S) 
    # equivalent to PCA.components_ in sklearn 
    components_ = v[idx,:] 
    return pcs,components_
```

This only works for the `arpack` solver. It's a bit slower than PCA on dense matrices (since arpack is slower than randomized), but it's super memory efficient.

**Pull request closed without merging**